File: 2021/11/15/main.go
package main

import "sort"

func largestDivisibleSubset(nums []int) []int {
	size := len(nums)
	if size == 0 {
		return []int{}
	}

	sort.Ints(nums)
	mx := 1
	idx := 0

	dp := make([]int, size)
	for i := range dp {
		dp[i] = 1
	}

	next := make([]int, size)

	for i := size - 2; 0 <= i; i-- {
		for j := size - 1; i < j; j-- {
			if nums[j]%nums[i] != 0 {
				continue
			}
			if dp[i] < dp[j]+1 {
				next[i] = j
				dp[i] = dp[j] + 1
			}
			if mx < dp[i] {
				mx = dp[i]
				idx = i
			}
		}
	}

	res := make([]int, mx)
	for i := range res {
		res[i] = nums[idx]
		idx = next[idx]
	}

	return res
}

File: 2021/11/15/main.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        if not nums: return []
        dp = [0]*len(nums)
        nums.sort()
        idx, size= 0, 0
        for i in range(len(nums)):
            k, c= -1, 0
            for j in range(i-1,-1,-1):
                if nums[i]%nums[j] == 0 and len(dp[j])>c:
                    c = len(dp[j])
                    k = j 
            if k!=-1:
                dp[i] = dp[k]+[nums[i]]
            else:
                dp[i] = [nums[i]]

            if len(dp[i])>size:
                idx = i
                size = len(dp[i])
        return dp[idx]